### PR TITLE
User quota support

### LINKED ISSUE #295: Implement support for user storage quotas
Currently, a user can full the /home and there is nothing an admin can do.

### PR DESCRIPTION
**Pull request by etiennedub**
close #295 
Depends on [magic_castle/282](https://github.com/ComputeCanada/magic_castle/pull/282)

Add support for user quota on NFS disk. The quota is the same for all users.

### CODE REVIEW

**Comment by cmd-ntrf** [comment]
I had no idea we could use xfs quotas over NFS, great find.

Questions:
- This sets a global quota for all users, can we think of a mechanism to specify a quota for a group of users?
- The exec has no parameter `onlyif`, `unless`, `create` or `refreshonly`, this means it will be executed at each Puppet apply. Is there a command to validate the quota has been properly configured and avoid re-executing it?
- Since quotas have no impact on infrastructure, I think we should avoid defining it in `main.tf` and instead make it a parameter provided via hieradata. Any objections?

**Comment by etiennedub** [comment]
It works over NFS because the permissions are preserved, so it's not limited to NFS. However, I haven't really found a way to do it with NFS directly.

> This sets a global quota for all users, can we think of a mechanism to specify a quota for a group of users?

Do we want to set a user quota for a specific set of users, or a share quota for a group of users (`grpquota`)? In both cases, I think the challenging part would be the Terraform configuration from a user perspective.

> The [exec](https://www.puppet.com/docs/puppet/7/types/exec.html#exec-attributes) has no parameter onlyif, unless, create or refreshonly, this means it will be executed at each Puppet apply. Is there a command to validate the quota has been properly configured and avoid re-executing it?

I think running it once would be the better option here. That way, an administrator could overwrite the quota if needed. I will fix it.

> Since quotas have no impact on infrastructure, I think we should avoid defining it in main.tf and instead make it a parameter provided via hieradata. Any objections?

I'm not sure how to pass them to hieradata without editing the infrastructure. Also, the volumes are defined in the infra module. Can you explain your suggestion?

**Comment by cmd-ntrf** [comment]
> It works over NFS because the permissions are preserved, so it's not limited to NFS. However, I haven't really found a way to do it with NFS directly.
> 
> > This sets a global quota for all users, can we think of a mechanism to specify a quota for a group of users?
> 
> Do we want to set a user quota for a specific set of users, or a share quota for a group of users (`grpquota`)? In both cases, I think the challenging part would be the Terraform configuration from a user perspective.

I envision it as a parameter of users definition as defined by 
https://github.com/computeCanada/puppet-magic_castle/?tab=readme-ov-file#profileusersldap

> 
> > The [exec](https://www.puppet.com/docs/puppet/7/types/exec.html#exec-attributes) has no parameter onlyif, unless, create or refreshonly, this means it will be executed at each Puppet apply. Is there a command to validate the quota has been properly configured and avoid re-executing it?
> 
> I think running it once would be the better option here. That way, an administrator could overwrite the quota if needed. I will fix it.

If the quota has to be adjusted, ideally, it should be via Puppet, not manually by the administrator.

> 
> > Since quotas have no impact on infrastructure, I think we should avoid defining it in main.tf and instead make it a parameter provided via hieradata. Any objections?
> 
> I'm not sure how to pass them to hieradata without editing the infrastructure. Also, the volumes are defined in the infra module. Can you explain your suggestion?

We need to define a new parameter in either profile::nfs::server or profile::users::ldap and have the MC admin provide the value for that parameter via the [hieradata YAML file](https://github.com/ComputeCanada/magic_castle/tree/main/docs#413-hieradata-optional).

**Comment by etiennedub** [comment]
I did update with the hieradata `profile::nfs::server::usrquotas` instead of adding an attribute to the volume and the exec is only executed when needed (refreshonly).

Also, I didn't really fix the problem for the specific user selection, but I changed the variable names to `usrquotas` instead of `quota`. That way, we can define `grpquotas` and set the user group with `profile::users::ldap`. However, I think this should be in another task and really define the use case that we need here. 

This make the Terraform MR useless, I will cancel it. Should I add something to the doc in MagicCastle (terraform)?

File: site/profile/manifests/nfs.pp [hunk 1]
```diff
@@ -125,10 +128,11 @@
...
   exec { "vgchange-${name}_vg":
-    command => "vgchange -ay ${name}_vg",
-    onlyif  => ["test ! -d /dev/${name}_vg", "vgscan -t | grep -q '${name}_vg'"],
-    require => [Package['lvm2']],
-    path    => ['/bin', '/usr/bin', '/sbin', '/usr/sbin'],
+    command     => "vgchange -ay ${name}_vg",
+    onlyif      => ["test ! -d /dev/${name}_vg", "vgscan -t | grep -q '${name}_vg'"],
+    require     => [Package['lvm2']],
+    path        => ['/bin', '/usr/bin', '/sbin', '/usr/sbin'],
+    refreshonly => true,
```
[the remark this answers is not in the document]
Comment (etiennedub) [reply site/profile/manifests/nfs.pp]: My bad... I was not suppose to be on this `exec`. It's fixed now

**Pull request merged**